health/health_json.c

```diff
@@ -1,92 +1,129 @@
-                    "\t\t\"status\": \"%s\",\n"
-                    "\t\t\"old_status\": \"%s\",\n"
-                    "\t\t\"delay\": %d,\n"
-                    "\t\t\"delay_up_to_timestamp\": %lu,\n"
-                    "\t\t\"updated_by_id\": %u,\n"
-                    "\t\t\"updates_id\": %u,\n"
-                    "\t\t\"value_string\": \"%s\",\n"
-                    "\t\t\"old_value_string\": \"%s\",\n"
-                    "\t\t\"last_repeat\": \"%lu\",\n"
-                    "\t\t\"silenced\": \"%s\",\n"
-                   , host->hostname
-                   , ae->unique_id
-                   , ae->alarm_id
-                   , ae->alarm_event_id
-                   , ae->name
-                   , ae->chart
-                   , ae->family
-                   , (ae->flags & HEALTH_ENTRY_FLAG_PROCESSED)?"true":"false"
-                   , (ae->flags & HEALTH_ENTRY_FLAG_UPDATED)?"true":"false"
-                   , (unsigned long)ae->exec_run_timestamp
-                   , (ae->flags & HEALTH_ENTRY_FLAG_EXEC_FAILED)?"true":"false"
-                   , ae->exec?ae->exec:host->health_default_exec
-                   , ae->recipient?ae->recipient:host->health_default_recipient
-                   , ae->exec_code
-                   , ae->source
-                   , ae->units?ae->units:""
-                   , (unsigned long)ae->when
-                   , (unsigned long)ae->duration
-                   , (unsigned long)ae->non_clear_duration
-                   , rrdcalc_status2string(ae->new_status)
-                   , rrdcalc_status2string(ae->old_status)
-                   , ae->delay
-                   , (unsigned long)ae->delay_up_to_timestamp
-                   , ae->updated_by_id
-                   , ae->updates_id
-                   , ae->new_value_string
-                   , ae->old_value_string
-                   , (unsigned long)ae->last_repeat
-                   , (ae->flags & HEALTH_ENTRY_FLAG_SILENCED)?"true":"false"
-    );
-
-    health_string2json(wb, "\t\t", "info", ae->info?ae->info:"", ",\n");
-
-    if(unlikely(ae->flags & HEALTH_ENTRY_FLAG_NO_CLEAR_NOTIFICATION)) {
-        buffer_strcat(wb, "\t\t\"no_clear_notification\": true,\n");
+static void
+health_kv_str(BUFFER *wb, const char *key, const char *value,
+       const char *prefix, const char *suffix) {
+    if (utf8_check(value)) {
+        error("Found malformed utf8 string for key %s", key);
+        value = "malformed_utf8_string";
     }
 
-    buffer_strcat(wb, "\t\t\"value\":");
+    buffer_sprintf(wb, "%s\"%s\": \"", prefix, key);
+    buffer_strcat(wb, value);
+    buffer_strcat(wb, "\"");
+    buffer_strcat(wb, suffix);
+}
+
+static void
+health_kv_lu(BUFFER *wb, const char *key, long unsigned value,
+      const char *prefix, const char *suffix) {
+    buffer_sprintf(wb, "%s\"%s\": %lu%s", prefix, key, value, suffix);
+}
+
+static void
+health_kv_ld(BUFFER *wb, const char *key, long int value,
```

Comment:
hmmm not sure this is the best way. this is defined in health_json.c but ideally we would have consistency with all the places where we generate JSON... what about chart/charts etc.